### ISSUE TITLE
Parallelise cabal build over modules

### ISSUE THREAD

**Issue by 23Skidoo** [haskell/cabal#976]
**Updated summary by @ezyang.** Previously, this ticket talked about all sorts of parallelism at many levels. Component-level parallelism was already addressed in #2623 (fixed by per-component builds), so all that remains is per-module parallelism. This is substantially more difficult, because right now we build by invoking `ghc --make`; achieving module parallelism would require teaching Cabal how to build using `ghc -c`. But this too has a hazard: if you don't have enough cores/have a serial dependency graph, `ghc -c` will be slower, because GHC spends more time reloading interface files. In https://github.com/haskell/cabal/issues/976#issuecomment-7016707 @dcoutts describes how to overcome this problem.

There are several phases to the problem:

1. First building the GHC build server and parallelism infrastructure. This can be done completely independently of Cabal: imagine a program which has a command line identical to GHC, but is internally implemented by spinning up multiple GHC processes and farming out the compilation process. You can tell if this was worthwhile when you get scaling better than GHC's built-in `-j` and a traditional `-c` setup.

2. Next, we need to teach Cabal/cabal-install how to take advantage of this functionality. If you implemented your driver program with exactly the same command line flags as GHC, then this is as simple as just passing `-w $your_parallel_ghc_impl`. However, this is a problem doing it this way: cabal-install will attempt to spin up N parallel package/component builds, which each in turn will try to spin up M GHC build servers; this is bad; you want the total number of GHC build servers to equal the number of cores.  So then you will need to setup some sort signalling mechanism to avoid too many build servers from running at once, OR have cabal new-build orchestrate the entire build down to the module level so it can plan parallelism (but you would probably have to rearchitect according to #4174 before you can do this.)

----

Now that the package-level parallel install has been implemented (see #440), the next logical step is to extend `cabal build` with support for building multiple modules, components and/or build variants (static/shared/profiling) in parallel. This functionality should be also integrated with `cabal install` in such a way that we don't over- or underutilise the available cores.

A prototype implementation of a parallel `cabal build` is [already available as a standalone tool](https://github.com/23Skidoo/ghc-parmake). It works by first extracting a module dependency graph with 'ghc -M' and then running multiple 'ghc -c' processes in parallel.

Since the parallel install code uses the external setup method exclusively, integrating parallel `cabal build` with parallel install will require using IPC. A single coordinating `cabal install -j N` process will spawn a number of `setup.exe build --semaphore=/path/to/semaphore` children, and each child will be building at most N modules simultaneously. An added benefit of this approach is that nothing special will have to be done to support custom setup scripts.

An important issue is that compiling with `ghc -c` is slow compared to `ghc --make` because the interface files are not cached. One way to fix this is to implement a "build server" mode for GHC. Instead of repeatedly running `ghc -c`, each build process will spawn at most N persistent ghcs and distribute the modules between them. Evan Laforge [has done some work in this direction](http://www.haskell.org/pipermail/glasgow-haskell-users/2012-February/021907.html).

Other issues:
- Building internal components in parallel requires knowing their dependency graph (this is being implemented as part of integrating `cabal repl` patches).
- Generating documentation in parallel may be only safe for `build-type: Simple`.

**Comment by angerman** [comment]
@23Skidoo what happened to this? Would this need to be restarted from scratch?

**Comment by 23Skidoo** [comment]
There's `ghc --make -j` now, which is used by `cabal build` automatically, though not by `cabal install -j`. Perhaps we should close this ticket and open a new one for making `cabal install -j` use `ghc --make -j`. #4174 is also relevant as a possible alternative to `--make -j`.

**Comment by angerman** [comment]
@23Skidoo is that also true for `new-build`?
I'm curious as `cabal new-build cabal-install --allow-newer` seems not to build in parallel.

**Comment by 23Skidoo** [comment]
I think it should. If you run it with `-v3`, it'll show you the ghc invocation string. You can also try `new-build -j`.

**Comment by 23Skidoo** [comment]
On a second thought, since `new-build` also installs dependencies, it probably suffers from the same problem as `install -j` and doesn't use `ghc --make -j`.

**Comment by angerman** [comment]
I didn't manage to go through the massive amount of output `-v3` produces. However looking at the processes, I see a single ghc process running. Where I'd like to see 4 or preferably even 8. Therefore I believe `cabal new-build -j` does not parallelize :-(

**Comment by 23Skidoo** [comment]
Yep, see my second comment. We should open a ticket for making `install/new-build -j` use `ghc --make -j`. I have some initial code on this branch: https://github.com/23Skidoo/cabal/commits/num-linker-jobs

**Comment by angerman** [comment]
@23Skidoo I'm perfectly fine with the idea of opening a second ticket, which details this; and closing this ticket. Please do! I just wanted to note down how this behaved on my system.

**Comment by nh2** [comment]
When doing so, just don't forget that `ghc --make -j` is horribly inefficient and needs RTS flags (like described on http://trofi.github.io/posts/193-scaling-ghc-make.html) to be of any use (otherwise it will be slower than non-j builds).

**Comment by angerman** [comment]
@nh2 so presumably we want 
```
ghc --make -j +RTS -A256M -qb0 -RTS
```

@trofi do you agree? Will this eventually be in GHC?

**Comment by ezyang** [comment]
So, based on this discussion, there seem to be two distinct issues:

1. My update summary, which is all about a fancy build server design to overcome the scaling problems of `ghc --make -j`
2. A much simpler fix, which is to make it so that you can pass `-j` to `ghc` when building things in parallel.

(2) should not be hard to implement, although it will be harder to demonstrate that it actually speeds things up (due to the fact that `ghc -j` scales pretty horribly). Perhaps we should make a separate ticket for it, though I don't think we should close this one.

**Comment by angerman** [comment]
If someone outlines how to do this proper, I might take a look at doing so.  My main quibble is that "GHC is slow" is a known theme. And if we aggravate this with cabal, we are potentially leaving build performance on the table.

**Comment by ezyang** [comment]
@angerman Well, which strategy did you want to do?

**Comment by angerman** [comment]
@ezyang what's the optimal solution we can hope for here? Turn cabal into using shake (#4174) and then see if how much of the additional build server we need?

@23Skidoo, if I understand correctly might have part of (2) already done?

**Comment by ezyang** [comment]
As I understand it, here is the optimal solution:

1. Implement GHC buildserver, with a GHC Shake driver that can submit requests for build to the buildserver
2. Rewrite Cabal in Shake, deferring to the GHC Shake driver for actually building Haskell code
3. Rewrite cabal-install in Shake, deferring to Cabal Shake driver for actually building a package/component

In the end you have a single Shake build system which knows about parallelism down to the individual Haskell source file level, and can do maximal parallelism.

...I don't expect this to be implemented any time soon.

**Comment by angerman** [comment]
I believe those three items can be built in separate steps with increasing productivity with each additional layer?

Could cabal produce a cmake (or make) file instead and feed that into build system?
I believe that's what ghc-cabal does to some extend?

**Comment by ezyang** [comment]
> I believe those three items can be built in separate steps with increasing productivity with each additional layer?

Yes.

> Could cabal produce a cmake (or make) file instead and feed that into build system?
I believe that's what ghc-cabal does to some extend?

Well, depends on what you mean by make. `ghc -M` knows how to create a Makefile for building Haskell. Cabal itself, in principle, knows about dependencies between components in a package, but this information isn't reified anywhere in the code today. Plus, a large amount of the processing is done by Haskell code in process, so your Makefile wouldn't have anything to call.

**Comment by angerman** [comment]
@ezyang 
Great! So one could start with the GHC build server for example and get somewhere.

What I'm wondering is if cabal, which has the notion of packages, targets and corresponding Modules/Files, (and flags), could generate a Makefile (or CMakeLists.txt -- assuming there was some plumbing for haskell), that exposed those targets.

And at least in the case of CMake, could be used to generate ninja files, which then in turn could be compiled using ninja as a build system, or even shake (which as far as I understand, can read ninja files).

This brings another question up. I believe there is some `ghc --make` powered by shake floating around somewhere, would investing to get that into ghc proper, help us with the `-j` scaling?

**Comment by ezyang** [comment]
> What I'm wondering is if cabal, which has the notion of packages, targets and corresponding Modules/Files, (and flags), could generate a Makefile (or CMakeLists.txt -- assuming there was some plumbing for haskell), that exposed those targets.

The big problem is that many operations which need to be done while building can't be characterized as just "run this and that command." There's a lot of Haskell code that gets run during a build, that needs to get run, and is not exposed as a command in any way. Shake has a similar problem: it's more expressive than ninja, so you can't take a Shake build system and turn it into ninja.

> I believe there is some ghc --make powered by shake floating around somewhere, would investing to get that into ghc proper, help us with the -j scaling?

There are two. You have https://github.com/ndmitchell/ghc-make which is implemented by calling `ghc -c` (you get parallelism but it is slower than `--make` sequentially), and https://github.com/ezyang/ghc-shake which uses the GHC API and is not parallelized.

Neither of these can be put into GHC because they depend on Shake and GHC does not want to take Shake on as a boot library at this time.

**Comment by nh2** [comment]
I am a bit confused by the mention of `ghc -M`. Is that even an option?

To my knowledge, `ghc -M` cannot reliably compile all Haskell projects in the same way as `ghc --make` can do.

None of [`ghc-make`](https://github.com/ndmitchell/ghc-make), [`ghc-parmake`](https://github.com/23Skidoo/ghc-parmake) can by themselves provide reliable compilation based on `ghc -M` as a Makefile is not expressive enough to do the up-to-date checks that `ghc --make` does.

---

Is this discussion still about how to combine package-level `-j` and module-level `-j`?

With the difficulty being that passing `-j4` to both would create 16 threads in the worst case and thus creating overhead?

If yes, implementing a ghc build server (while useful) to solve this problem sounds _way_ out of scope to get some basic solution going in the short term.

Here's a commonly used alternative: `make` supports two flags, `-j` and `-l`. The `-l` flag (`--load-average`)

> Specifies  that  no  new jobs (commands) should be started if there are others jobs running and the load average is at least load (a floating-point number).  With no argument, removes a previous load limit.

Many build systems that call `make` recursively use this functionality to solve the same problem we have here; they use `make -j4 -l4`, to ensure that globally no more than roughly 4 CPU cores are fully used.

How about an `-l`-style flag for GHC? Then cabal can always call `ghc -j` or `ghc -j4` while also compiling multiple packages itself. This would even be maximally efficient, because cabal-level `-j` is more efficient than ghc-level `-j`, and a `make -l` style approach would always result in full cabal-level parallelisation wen parallelisation is possible at both levels.

**Comment by Blaisorblade** [comment]
@angerman 
> The big problem is that many operations which need to be done while building can't be characterized as just "run this and that command." 

There are also two big reasons why build steps aren't worth exposing.
- if your (c)make-based build has a GHC process for source file, each GHC process will spend time to reload interfaces written by other files, as mentioned.
- Worse, if Foo.hs changes, but the new Foo.hi has the same content, you don't want to rebuild Bar.hs. At least with Make, that's extra logic to add to replace the *base* building block of Makefiles.

**Comment by nh2** [comment]
I forgot to mention in https://github.com/haskell/cabal/issues/976#issuecomment-298360770:

The approach I described is exactly how `nix` avoids the same problem as we have here: `nix` itself can build multiple packages (say C projects) at once, and each package itself can be compiled with `make -j`.

**Comment by Blaisorblade** [comment]
@nh2 Didn't know about `-l` sounds very interesting, and could be easier to implement (at least on *x)!
But one still risks starting too many long-lived tasks at once, when the load average is still low. Still, it might be worth studying its performance in practice. Is there any benchmark showing speedups in the Nix scenario? It's not the exact same load, but it's already implemented :-)

**Comment by nh2** [comment]
@Blaisorblade The tasks should not be long-lived (at the module-level); when implementing `-l`, ghc should check the load every time a module finishes or wants to start. Most operating systems can do very well when the number of tasks is higher than the number of processes; much better than `ghc` can indeed. The only time it is a problem is when you are very memory constrained, e.g. when even temporarily compiling 16 modules (over 4 processes) immediately makes you go out of memory. If you have such a system, you would most likely not want full `-j` anyway; if you do, there are other mitigations, such as waiting a short bit before starting the next job in order to get an accurate load number.

In any case, the non-Haskell world seems to have done pretty well with `-l` (although in general it also doesn't have compilers that take GBs of memory; with the exception of C++ maybe).

**Comment by ezyang** [comment]
> None of ghc-make, ghc-parmake can by themselves provide reliable compilation based on ghc -M as a Makefile is not expressive enough to do the up-to-date checks that ghc --make does.

I'll remark that ghc-shake, by virtue of using the GHC API, has as precise recompilation checking as `ghc --make`.

`-l` seems like a reasonable idea, assuming that you don't mind this new feature not working unless you're using GHC 8.4 ;)

**Comment by Blaisorblade** [comment]
Operationally, it's enough if `-l` is convincing enough for somebody to volunteer implementing it (sadly, not me) ;-) I'm still 👍. https://www.preney.ca/paul/archives/341 also recommends it for Gentoo...

Anyway, while we're trying to anticipate how good it's going to be:

> The tasks should not be long-lived (at the module-level); when implementing -l, ghc should check the load every time a module finishes or wants to start.

I meant that some modules take longer to build...

> The only time it is a problem is when you are very memory constrained

You also had IO, but with SSDs we can probably ignore that. Based on e.g. http://stackoverflow.com/a/17749621/53974 you're essentially right—build times increase only a tiny bit for higher `-j`.

OTOH, I'd guess we *are* memory constrained when you rebuild ~50 packages (say, a fresh Agda build).

> In any case, the non-Haskell world seems to have done pretty well with -l (although in general it also doesn't have compilers that take GBs of memory; with the exception of C++ maybe).

IIRC C++ has also files taking longer to build, so maybe C++ builds could be a better benchmark than C, if somebody looks for results/tries it out...

**Comment by nh2** [comment]
> OTOH, I'd guess we are memory constrained when you rebuild ~50 packages (say, a fresh Agda build).

@Blaisorblade I don't get it, does the number of packages matter?

Just to be explicit, the way I imagine it to work is this:

* Cabal sees you have 4 cores.
* Cabal starts building 4 packages in parallel.
* So Cabal invokes `ghc -j4 -l4`, 4 times.
* Each of the ghcs checks the current load before/after every build step (module).
* Thus in the worst case, you are initially compiling 16 modules at the same time.
* As soon as the some ghc completes building its first module, it'll see that the load is > 4 and spawn no replacement for the completed module.
* As a result, the number of modules compiled globally quickly converges from 16 to 4; as soon as the 17th overall module starts compilation, we can be quite sure that there are only 4 modules building in total.

This seems to be the case no matter if you have 4 packages or 50.

**Comment by Blaisorblade** [comment]
You're right, 50 packages shouldn't affect the number of processes. Need more coffee.
Regarding load average: on Linux `uptime` gives the one-minute average, so you'd have some delay (1/n minutes, where n is your example). Or do you have some other stat in mind?
http://stackoverflow.com/a/21621256/53974

**Comment by nh2** [comment]
It seems that GNU make indeed uses the 1-minute load average:

http://git.savannah.gnu.org/cgit/make.git/tree/getloadavg.c?h=4.2.1#n491

```c
/* Put the 1 minute, 5 minute and 15 minute load averages
   into the first NELEM elements of LOADAVG.
   Return the number written (never more than 3, but may be less than NELEM),
   or -1 if an error occurred.  */

int
getloadavg (double loadavg[], int nelem)
{
```

Used here: http://git.savannah.gnu.org/cgit/make.git/tree/job.c?h=4.2.1#n1945

However, since Linux's load average is an exponentially weighted average instead of a real average, this might still be a non-issue. For example, `stress-ng --cpu 16` on my 4-core machine brings the 1-minute load time in `htop` above for after already 3 seconds or so.

**Comment by Blaisorblade** [comment]
> However, since Linux's load average is an exponentially weighted average instead of a real average, this might still be a non-issue.

Good point, didn't know that. 👍 Googling leads to http://juliano.info/en/Blog:Memory_Leak/Understanding_the_Linux_load_average confirming and elaborating.

To sum up: what you proposed thinking is close to `-j16 -l 4` rather than  `-j4 -l 4`. You gave reasons to believe it should work, including prior art from Nix. We've debated potential costs and this still seems a very promising idea. Who'd volunteer to champion this in GHC?

**Comment by angerman** [comment]
> To sum up: what you proposed thinking is close to -j16 -l 4 rather than -j4 -l 4. You gave reasons to believe it should work, including prior art from Nix. We've debated potential costs and this still seems a very promising idea. Who'd volunteer to champion this in GHC?

Unless someone else feels like it, I can give this a shot. Seems like it shouldn't be too hard. But who knows.

**Comment by angerman** [comment]
Looking at that `getloadavg` function in make, does anyone know of a library that provides those information sufficiently enough for ghc's use case? I'd rather not copy those 500 lines of code, if there is a simpler way.

**Comment by Blaisorblade** [comment]
1. This code is probably what you want (found through Hayoo). You might need to copy-paste it eventually
http://hackage.haskell.org/package/loadavg-0.1/docs/System-Posix-LoadAvg.html

For bonus points, it uses `getloadavg` (see `man 3 getloadavg`), which has the same API on Linux and OS X (and comes from BSD 4.3 so should work on other BSDs).

Those 500 lines for Make support more obscure platforms (half of which aren't supported by GHC anyway), and seem to not support either OS X or Windows anyway.

**Comment by angerman** [comment]
Upon a suggestion by @ezyang, I tried making `ghc` be invoked with `-j`.

For the record: timings by just forwarding `-j` to ghc, when building `cabal-install`

```
cabal new-build cabal-install --allow-newer  780.90s user 12.66s system 98% cpu 13:27.41 total
cabal new-build cabal-install --allow-newer -j  784.95s user 14.14s system 98% cpu 13:31.16 total
cabal new-build cabal-install --allow-newer -j --ghc-options=-j  1216.44s user 265.02s system 359% cpu 6:52.49 total
```

**Comment by angerman** [comment]
After giving this some thought and looking into ghc. I wonder if we want `cabal -l` rather than `ghc -l`. Compared with make, that would be where `-l` is passed as well. We don't pass `-l` to the downstream compilers usually.

@hvr suggested `-j <n>:<m>` for cabal, were `<n>` is package level, and `<m>` is ghc-level and passed to `ghc -j`.

Thus, I'd suggest the following:
- Add `-l` to cabal
- Allow `-j <n>:<m>`.

There still the question how to handle `-j <n>` if we allow it. Would it be `-j <n>:<n>` or `-j <n>:<1>` or `-j <1>:<n>`?

**Comment by nh2** [comment]
> After giving this some thought and looking into ghc. I wonder if we want cabal -l rather than ghc -l. Compared with make, that would be where -l is passed as well. We don't pass -l to the downstream compilers usually.

@angerman I don't understand this. Sure you wouldn't pass `-l` to `gcc`. But we wouldn't pass `-l` to ghc-the-compiler, we would pass `-l` to `ghc --make`-the-parallelizing-build-system.

The lowest level component that spawn threads must have the `-l` option; in our case that's `ghc --make`.

Of course Cabal can have a user-exposed `-l` option as well in addition, but without `ghc` having it, it can't work.

**Comment by angerman** [comment]
After discussing this with @ezyang in #hackage for a bit, I don't think I'm going to try and vedge `-l` into ghc. My understanding is that ghc starts with `<n>` capabilities and then distributes work among them. Where work is basically a thread per module, and the the threads form a dependency graph. Thus if we wanted to check against the load average before starting a new thread (or dynamically shrinking / widening the capabilities), I don't see a trivial solution to this; and I don't intend to layer much more complexity onto ghc, as this would be a stop gap measure until we `ghcd`, as @ezyang suggested. I'd argue that time is better spent on `ghcd` than trying to wedge `-l` into ghc.

If we started to check again `-l` at the beginning when setting up capabilities, this would be identical to adding `-l` to cabal as far as I can see. If this renders no benefits as @nh2 points out, we might just have to accept that `-l` is not an easy option with the way `ghc --make` is designed. 

I'm open to alternative ideas!

**Comment by nh2** [comment]
If there are `n` threads (e.g. `forkIO`) set up at the beginning, the `-l` concept should still work.

There will be some mechanism in the build system that determines when the next module can start to build.

I imagine the load check should simply go into that place (as opposed to grow or shrink the number of threads, the threads would still be there, but be blocked).

**Comment by treeowl** [comment]
Is the speed at which GHC can read interface files a bottleneck? How hard might it be to fix that?

**Comment by nh2** [comment]
> Is the speed at which GHC can read interface files a bottleneck?

@treeowl I'm not sure if that is known.

I've advertised in some other place that GHC, for the various parts of its build pipeline, should record CPU and wall time and be able to produce a report (e.g. "N seconds CPU/wall were spent on reading and decoding interface files). That way we could more easily pinpoint where bottlenecks are. Right now GHC does things time counting and reporting only for optimiser phases, not for any of the "more basic tech" bits.

**Comment by duog** [comment]
Hi All,

I've written a prototype for a ghc feature to limit its parallelism with a semaphore, GNU make jobserver style. The idea being that cabal-install would pass `-j -jsem ghc` to `ghc --make` on each invocation. There is a datum showing a nice speedup on building lens from scratch.  

https://gitlab.haskell.org/ghc/ghc/-/merge_requests/5176 

I'd appreciate any comment on
* why this is a bad idea
* what the UI should look like
* any important semantics 

to make this easy and useful to integrate into cabal

**Comment by Profpatsch** [comment]
fwiw, I was able to fix that for local development in our project by adding

```
package <projectname>
  -- This causes cabal to build modules on all cores, instead of just one.
  ghc-options: -j
```

to my `cabal.project`.

This should be the default for interactive development, and I was kinda shocked that it isn’t!

**Comment by Mikolaj** [comment]
@Profpatsch: Great! What are your precise results?

**Comment by Martinsos** [comment]
@Profpatsch thanks for the tip, it also sped up things for me!

@Mikolaj , so I first added
```
jobs: $ncpus
```
thinking that will speed up local development / building of local code.

It however had no effect, so I started searching through cabal issues on GH and found this issue. I applied the suggestion by @Profpatsch : 
```
package waspc
  -- This causes cabal to build modules on all cores, instead of just one,
  -- therefore reducing our build times.
  ghc-options: -j
```

This reduced build time of our project from 43 seconds to 26 seconds, which is pretty nice!

I ran measurements with `cabal clean && time cabal build`.

I also wonder why this is not default, I guess it is relatively new thing?
It might also be valuable explaining in the docs that `jobs: $ncpus` parallelizes only on the package level, not module level, and in the meantime while it is don't a default thing, mention this trick with `ghc-options: -j`.

**Comment by Mikolaj** [comment]
Thank you for the data. Good points. I wonder how reporting warning and errors works with that option? E.g., may two warning texts be interspersed? I think that was in the past the blocker for cabal-level `-j` but it's been somehow taken care of (I don't know the details).

**Comment by Mikolaj** [comment]
BTW, there is active work by some GHC hacker on how to allocated cores to cabal-level and GHC-level `-j`. I see in your case there's speedup with a default allocation, but in general, you can get a slowdown due to cores being stolen from cabal by GHC or the other way around.

**Comment by Mikolaj** [comment]
So, I guess, default GHC `-j` could potentially slow down initial builds of packages, but speed up subsequent builds. Hard to balance automatically...

**Comment by Martinsos** [comment]
@Mikolaj thanks for explaining!

I have to admit I don't know how errors are reported, haven't tried that out.

So help me understand if I got this right: if cabal is building only one package, then GHC `-j` can be only beneficial. But, if cabal is building multiple packages and trying to parallelize that, then GHC `-j` can mess that up because it is also trying to parallelize stuff on its own, and then both of them trying to parallelize stuff ends up with them fighting for resources (threads)?

But I guess with the way I set it up now, that shouldn't be a problem, because GHC `-j` is enabled only for my local package, and not for the external packages -> so when installing external dependencies, there is only one parallelization happening, and that is on the package level (by cabal), while when building my local project (after external dependencies have been installed), only GHC will be doing parallelization?

**Comment by fgaz** [comment]
RAM also fills up quickly with -j

**Comment by Martinsos** [comment]
I understand then that having GHC `-j` enabled for libraries would not be great, at least not when they are built as external dependencies.

But it still sounds useful for local development of a package, be it library or executable.

So maybe it makes sense for most people to have GHC `-j` specified in cabal.project.local? Is that something that could be recommended by default? Or is that also situational?

**Comment by Mikolaj** [comment]
Yes, I think GHC `-j` is a good default for `cabal.project.local`. Not always the best, e.g., when a project has many packages or a package has many components that the user wants to always build (then they are built in parallel), but these are less common situations.

**Comment by Martinsos** [comment]
Regarding ensuring that freeze file is up to date: what I just did in our project was add this check in the CI that checks if freeze file is up to date, so I will share it here in case it is interesting for the discussion.
If freeze file is not up to date, CI fails. So although there is no automatic updating of freeze file by cabal, this ensures that it is indeed up to date. It is a stronger mechanism then automatic updating by cabal, but on the other hand most people probably won't think of doing it hm.

PR: https://github.com/wasp-lang/wasp/pull/508/files

Crucial code (Github Actions):
```
      - name: Check if cabal freeze file is up to date
        if: matrix.os == 'ubuntu-latest'
        run: |
          # We do the check by running `cabal freeze` and checking if cabal.project.freeze
          # changed. If it didn't, it is up to date, otherwise it is not.
          # If there is no cabal.project.freeze, or `cabal freeze` command failed, that is also red flag.
          [[ -f cabal.project.freeze ]] || exit 1
          OLD_FREEZE_SUM=$(md5sum cabal.project.freeze)
          cabal freeze || exit 1
          NEW_FREEZE_SUM=$(md5sum cabal.project.freeze)
          [[ "$NEW_FREEZE_SUM" == "$OLD_FREEZE_SUM" ]]
```

One thing I believe I learned from this, while thinking about regenerating freeze file, is that in most cases, the best thing to do is to first delete freeze file and then run `cabal freeze`. This might result in multiple dependencies changing, but is also guaranteed to not cause any unneeccesary complications. While on the other hand, if you run `cabal freeze` without deleting the freeze file first (for example in case you added a new dependency), there is a chance dependency resolution will fail due to all the versions that are pinned down by current freeze file.

So if cabal was to implement automatic regeneration of freeze file, it would almost certainly have to consist of first deleting existing freeze file and only then generating a new one (or using some other way to generate a new one while ignoring the existing one). I believe npm uses the same approach when regenerating package.lock.json -> it ignores existing one while doing it, and doesn't just add to it, but regenerates it whole, as if it never existed.

**Comment by Profpatsch** [comment]
> when a project has many packages or a package has many components that the user wants to always build (then they are built in parallel)

Cabal packages are a horror story anyway, e.g. they don’t work together with `cabal repl` very well (`:r` doesn’t work over package boundaries).

For local development I always want to use all available cores, for building libraries we use nixpkgs/nix, which knows how to forward the right amount of cores to its builds.



> this reduced our building time from 43s to 26s.
> When building the whole project (library + exe + tests) (`cabal clean && time cabal build --enable-tests`), this reduced our building time from 56s to 42s.

You can probably get more speedups if your project’s modules are split up in a reasonable way, i.e. there is no bottleneck module that all of compilation has to wait on. `Types.hs` comes to mind as the biggest antipattern.

**Comment by Profpatsch** [comment]
In order to see the module dependency graph, I use a script like this in our production code: 

```nix
  # display a graph of all modules and how they depend on each other
  mainserv-module-deps-with-filetype = self.writers.writeBash "mainserv-module-deps-with-filetype" ''
    shopt -s globstar
    filetype="$1"
    ${self.haskellPackages.graphmod}/bin/graphmod \
      ${/*silence warnings for missing external dependencies*/""} \
      --quiet \
      ${/*applies some kind of import simplification*/""} \
      --prune-edges \
      ${self.mainserv-root-directory}/src/**/*.hs \
      | ${self.graphviz}/bin/dot \
          ${/*otherwise it’s a bit cramped*/""} \
          -Gsize="20,20!" \
          -T"$filetype"
  '';
```

Which uses the very good https://hackage.haskell.org/package/graphmod command.

Then it’s just a matter of looking at the graph and noticing bottlenecks.

**Comment by Profpatsch** [comment]
The output for `graphmod --quiet --prune-edges waspc/src/**/*.hs | dot -Gsize=20,20! -Tpng` for example:

![image](https://user-images.githubusercontent.com/3153638/158808194-f7cafba9-4e45-465c-996c-f8da857c15ae.png)

**Comment by fgaz** [comment]
ghc-proposals/ghc-proposals#540 is in! :rocket:
